File: Medium/34.py
class Solution:
    def lowerBound(self, nums: List[int], target: int) -> int:
        low = 0
        up = len(nums)
        while low < up:
            mid = (low + up) // 2
            if nums[mid] >= target:
                up = mid
            else:
                low = mid + 1
        return low
    
    def upperBound(self, nums: List[int], target: int) -> int:
        low = 0
        up = len(nums)
        while low < up:
            mid = (low + up) // 2
            if nums[mid] > target:
                up = mid
            else:
                low = mid + 1
        return low

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return [-1, -1]
        else:
            low = self.lowerBound(nums, target)
            up = self.upperBound(nums, target) - 1
            if low == len(nums) or nums[low] != target:
                return [-1, -1]
            return [low, up]        

File: Medium/34/34.java
class Solution {
    private int findLeft(int[] nums, int target) {
        int i = 0;
        int j = nums.length;
        while (i < j) {
            int mid = i + (j - i) / 2;
            if (nums[mid] >= target) {
                j = mid;
            } else {
                i = mid + 1;
            }
        }
        return nums[i] == target ? i : -1;
    }

    private int findRight(int[] nums, int target) {
        int i = 0;
        int j = nums.length;
        while (i < j) {
            int mid = i + (j - i) / 2;
            if (nums[mid] <= target) {
                i = mid + 1;
            } else {
                j = mid;
            }
        }
        int res = i - 1;
        return nums[res] == target ? res : -1;
    }

    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0 || target < nums[0] || target > nums[nums.length - 1]) {
            return new int[] {-1, -1};
        }
        int[] res = new int[2];
        res[0] = findLeft(nums, target);
        res[1] = findRight(nums, target);
        return res;
    } 
}